clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -0,0 +1,435 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.utils.ByteBufferInputStream;
+import org.apache.kafka.common.utils.ByteUtils;
+import org.apache.kafka.common.utils.Crc32;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.apache.kafka.common.record.Records.LOG_OVERHEAD;
+
+/**
+ * RecordBatch implementation for magic 2 and above. The schema is given below:
+ *
+ * RecordBatch =>
+ *  BaseOffset => Int64
+ *  Length => Int32
+ *  PartitionLeaderEpoch => Int32
+ *  Magic => Int8
+ *  CRC => Uint32
+ *  Attributes => Int16
+ *  LastOffsetDelta => Int32
+ *  BaseTimestamp => Int64
+ *  MaxTimestamp => Int64
+ *  ProducerId => Int64
+ *  ProducerEpoch => Int16
+ *  BaseSequence => Int32
+ *  Records => [Record]
+ *
+ * Note that when compression is enabled (see attributes below), the compressed record data is serialized
+ * directly following the count of the number of records.
+ *
+ * The CRC covers the data from the attributes to the end of the batch. Note that the location is
```

Comment:
It's true that the change is not strictly needed. The reasoning for the switch is that this way the CRC includes everything that follows it (same as for V0 and V1 formats). In V0 and V1, the offset and length are before the CRC and are not included in the computation. In V2, the offset, length, partitionLeaderEpoch and magic are before the CRC and not included in the computation.

Given that, do you think the benefit is worth the switch?